pipeline/processor/postprocessor.py

```diff
@@ -0,0 +1,268 @@
+            .sort_values(["page_num", "block_num"])
+            .drop(columns=["page_num", "block_num"])
+            .to_dict("records")
+        )
+        return new_text_blocks
+
+    def _create_custom_attributes(self, blocks: List[Dict]) -> dict:
+        """
+        Create custom attributes for a list of text blocks.
+
+        Args:
+            blocks: List of text blocks belonging to a list element.
+
+        Returns:
+            A dictionary with appended custom attributes for a list element.
+        """
+
+        df = pd.DataFrame(blocks)
+        # Note this drop duplicates step is a bit of a hack to deal with undiagnosed upstream
+        # issues leading to duplicates..
+        df = df.loc[df.astype(str).drop_duplicates().index]
+        df["page_num"] = (
+            df["text_block_id"].str.split("_").str[0].str.extract("(\d+)").astype(int)
+        )
+
+        full_list_text = df["text"].tolist()
+        paths = df["path"].tolist()
+        block_ids = df["text_block_id"].tolist()
+        custom_bounding_boxes = (
+            df.groupby("page_num")
+            .apply(lambda x: self._minimal_bounding_box(x["coords"]))
+            .tolist()
+        )
+        custom_attributes_dict = {
+            "paths": paths,
+            "text_block_ids": block_ids,
+            "custom_bounding_boxes": custom_bounding_boxes,
+            "pretty_list_string": self._pprint_list(df),
+        }
+        new_dict = {
+            "text_block_id": block_ids[0],
+            "type": "list",
+            "text": full_list_text,
+            "custom_attributes": custom_attributes_dict,
+        }
+        return new_dict
+
+    def _group_list_elements(self, contents: dict, filename) -> dict:
+        """
+        Parse Adobe outputs to group list elements
+
+        Args:
+            contents: A dict of the processed adobe output for a particular pdf.
+
+        Returns:
+            Reprocessed version of the Adobe output with grouped list elements.
+        """
+
+        prev_list_ix = 0
+        blocks_seen = 0
+        last_page_ix = 0
+        new_pages = []
+        for ix, page in enumerate(contents["pages"]):
+            previous_block = None
+            new_text_blocks = []
+            dd = defaultdict(list)
+            for text_block in page["text_blocks"]:
+                blocks_seen += 1
+                if text_block["path"]:
+                    list_group = self._find_first_occurrence(
+                        self.regex_pattern, text_block["path"]
+                    )
+                    if list_group:
+                        current_list_id = f"{ix}_{list_group}"
+                        # Handle the case where we have a new list at the beginning of a page and where
+                        # the previous list block is assumed context.
+                        if (
+                            text_block["text_block_id"].split("_")[1] == "b1"
+                        ) and previous_block:
```

Comment:
I don't understand how this test works; if the text block we're currently holding is the first element on a page, aren't we guaranteed that `previous_block` is `None`?